### PR TITLE
Update full dreambooth script to work with IF

### PR DESCRIPTION
**Pull request by williamberman**
This updates the full dreambooth script to work with IF

To train the XL IF model using 8bit adam, you need ~48 GB of VRAM.

I also found it required very low learning rates or model quality would degrade.

Training command
```sh
export MODEL_NAME="DeepFloyd/IF-I-XL-v1.0"

export INSTANCE_DIR="dog"
export OUTPUT_DIR="dreambooth_if"

accelerate launch train_dreambooth.py \
  --report_to="wandb" \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --instance_data_dir=$INSTANCE_DIR \
  --output_dir=$OUTPUT_DIR \
  --instance_prompt="a photo of sks dog" \
  --resolution=64 \
  --train_batch_size=4 \
  --gradient_accumulation_steps=1 \
  --learning_rate=1e-7 \
  --max_train_steps=150 \
  --validation_prompt "a photo of sks dog" \
  --validation_steps 25 \
  --text_encoder_use_attention_mask \
  --tokenizer_max_length 77 \
  --pre_compute_text_embeddings \
  --use_8bit_adam \
  --set_grads_to_none \
  --skip_save_text_encoder
```

Training run: https://wandb.ai/williamberman/dreambooth/runs/wwmjvj83/

Running the model

```py
from diffusers import DiffusionPipeline
from transformers import T5EncoderModel

text_encoder = T5EncoderModel.from_pretrained("DeepFloyd/IF-I-XL-v1.0", subfolder="text_encoder")

pipe = DiffusionPipeline.from_pretrained("williamberman/dreambooth_if_dog", text_encoder=text_encoder)

pipe.to('cuda')

pipe('sks dog').images[0].save('out_1.png')

# Note that this shows a generic dog and not the concept we're training on. This
# isn't enough to not merge but we should look at if we use this more extensively
pipe('sks dog on a boat').images[0].save('out_2.png') 

pipe('a horse').images[0].save('out_3.png')
```
out_3
out_2
out_1

### CODE REVIEW

File: src/diffusers/models/unet_2d_blocks.py [hunk 1]
```diff
@@ -1478,16 +1478,33 @@ def forward(
         cross_attention_kwargs = cross_attention_kwargs if cross_attention_kwargs is not None else {}
 
         for resnet, attn in zip(self.resnets, self.attentions):
-            # resnet
```
Comment (williamberman) [inline src/diffusers/models/unet_2d_blocks.py:1481]: Adding gradient checkpointing for the other two IF unet blocks

File: examples/test_examples.py [hunk 2]
```diff
@@ -147,6 +147,32 @@ def test_dreambooth(self):
...
+                --output_dir {tmpdir}
+                --pre_compute_text_embeddings
+                --tokenizer_max_length=77
+                --text_encoder_use_attention_mask
+                """.split()
+
+            run_command(self._launch_args + test_args)
+            # save_pretrained smoke test
+            self.assertTrue(os.path.isfile(os.path.join(tmpdir, "unet", "diffusion_pytorch_model.bin")))
+            self.assertTrue(os.path.isfile(os.path.join(tmpdir, "scheduler", "scheduler_config.json")))
```
Comment (sayakpaul) [inline examples/test_examples.py:174]: Nice job adding this since the scheduler is overwritten 👍

File: examples/dreambooth/train_dreambooth.py [hunk 3]
```diff
@@ -608,6 +690,50 @@ def __getitem__(self, index):
...
+        return_tensors="pt",
+    )
+
+    return text_inputs
+
+
+def encode_prompt(text_encoder, input_ids, attention_mask, text_encoder_use_attention_mask=None):
+    text_input_ids = input_ids.to(text_encoder.device)
+
+    if text_encoder_use_attention_mask:
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth.py:723]: Did you notice any timing improvement with and without this?

File: src/diffusers/models/unet_2d_blocks.py [hunk 4]
```diff
@@ -1478,16 +1478,33 @@ def forward(
...
-            hidden_states = resnet(hidden_states, temb)
+            if self.training and self.gradient_checkpointing:
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d_blocks.py:1481]: Worth some comments before this branch.

**Review by sayakpaul — APPROVED** [review]
Good for me except a few comments 👍

File: examples/dreambooth/train_dreambooth.py [hunk 5]
```diff
@@ -961,37 +1139,48 @@ def load_model_hook(models, input_dir):
...
 
-                # Add noise to the latents according to the noise magnitude at each timestep
+                # Add noise to the model input according to the noise magnitude at each timestep
                 # (this is the forward diffusion process)
-                noisy_latents = noise_scheduler.add_noise(latents, noise, timesteps)
+                noisy_model_input = noise_scheduler.add_noise(model_input, noise, timesteps)
 
                 # Get the text embedding for conditioning
-                encoder_hidden_states = text_encoder(batch["input_ids"])[0]
+                # encoder_hidden_states = text_encoder(batch["input_ids"])[0]
```
Comment (stefsietz) [inline examples/dreambooth/train_dreambooth.py:1170]: Hey, just looked into this PR to understand IF training. This part is confusing in my opinion, and potentially buggy? "input_ids" refers to token ids afaik, which makes sense. If its the pre-encoded prompt hidden states, maybe put them under a different key into the batch?
Also, I think without a check whether it's token ids or encoded hidden state, this breaks the case where it's token ids.

File: examples/dreambooth/train_dreambooth.py (hunk 3 above)
Comment (williamberman) [reply examples/dreambooth/train_dreambooth.py:723]: I wasn't measuring any perf for the attention mask, this was just to match how the IF pipeline encodes text :)

File: src/diffusers/models/unet_2d_blocks.py (hunk 1 above)
Comment (williamberman) [reply src/diffusers/models/unet_2d_blocks.py:1481]: Note that this means we were still doing gradient checkpointing previously for IF but just making less checkpoints

Comment (sayakpaul) [reply src/diffusers/models/unet_2d_blocks.py:1481]: For curiosity: in the earlier case which blocks were not being chrckpointed?

File: examples/dreambooth/train_dreambooth.py [hunk 6]
```diff
@@ -835,6 +969,44 @@ def load_model_hook(models, input_dir):
...
+    if args.pre_compute_text_embeddings:
+
+        def compute_text_embeddings(prompt):
+            with torch.no_grad():
+                text_inputs = tokenize_prompt(tokenizer, prompt, tokenizer_max_length=args.tokenizer_max_length)
+                prompt_embeds = encode_prompt(
+                    text_encoder,
+                    text_inputs.input_ids,
+                    text_inputs.attention_mask,
+                    text_encoder_use_attention_mask=args.text_encoder_use_attention_mask,
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth.py:981]: I don't think `T5EncoderModel` uses `text_encoder_use_attention_mask` no?

**Comment by patrickvonplaten** [comment]
Good for me to merge! Let's make sure though that we add a small "How-to" section to the dreambooth README both for LoRA and for Dreambooth LoRA

File: src/diffusers/models/unet_2d_blocks.py (hunk 1 above)
Comment (williamberman) [reply src/diffusers/models/unet_2d_blocks.py:1481]: SimpleCrossAttn{Down,Up}Block2D :)
